src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -15336,6 +15336,143 @@ struct SliceExtend final : OpRewritePattern<enzymexla::ExtendOp> {
+
+    Value triggerOperand = triggerWrapOp.getOperand();
+    auto triggerSliceOp = triggerOperand.getDefiningOp<stablehlo::SliceOp>();
+
+    if (!triggerSliceOp) {
+      return failure();
+    }
+
+    Value baseOperand = triggerSliceOp.getOperand();
+
+    // --- Get Target Wrap Parameters ---
+    int64_t targetWrapDim = triggerWrapOp.getDimension();
+    int64_t targetLhs = triggerWrapOp.getLhs();
+    int64_t targetRhs = triggerWrapOp.getRhs();
+    Location loc = triggerWrapOp.getLoc();
+
+    // --- Check Validity of Trigger Slice ---
+    auto baseOperandType = dyn_cast<RankedTensorType>(baseOperand.getType());
+    if (!baseOperandType || !baseOperandType.hasStaticShape()) {
+      return rewriter.notifyMatchFailure(triggerWrapOp,
+                                         "Base operand requires static shape");
+    }
+    if (triggerSliceOp.getStartIndices()[targetWrapDim] != 0 ||
+        triggerSliceOp.getLimitIndices()[targetWrapDim] !=
+            baseOperandType.getShape()[targetWrapDim] ||
+        triggerSliceOp.getStrides()[targetWrapDim] != 1) {
+      return rewriter.notifyMatchFailure(
+          triggerWrapOp,
+          "Trigger SliceOp modifies the dimension being wrapped");
+    }
+
+    llvm::SmallVector<CandidateInfo> candidates;
+    candidates.push_back({triggerWrapOp, triggerSliceOp});
+
+    for (auto const &userOp : baseOperand.getUsers()) {
+      // Skip the slice that defines the trigger operand
+      if (userOp == triggerSliceOp.getOperation())
+        continue;
+
+      // Case 1: Direct Wrap
+      if (auto directWrap = dyn_cast<enzymexla::WrapOp>(userOp)) {
+        if (directWrap.getDimension() == targetWrapDim &&
+            directWrap.getLhs() == targetLhs &&
+            directWrap.getRhs() == targetRhs) {
+          candidates.push_back({directWrap, nullptr});
+        }
+      }
+      // Case 2: Wrap of Slice
+      else if (auto sliceUser = dyn_cast<stablehlo::SliceOp>(userOp)) {
+        if (!sliceUser->hasOneUse())
+          continue;
+        auto wrapOfSlice =
+            dyn_cast<enzymexla::WrapOp>(*sliceUser->user_begin());
+        if (!wrapOfSlice)
+          continue;
+
+        if (wrapOfSlice.getDimension() == targetWrapDim &&
+            wrapOfSlice.getLhs() == targetLhs &&
+            wrapOfSlice.getRhs() == targetRhs) {
+          // Check validity: sliceUser must not modify targetWrapDim
+          if (sliceUser.getStartIndices()[targetWrapDim] == 0 &&
+              sliceUser.getLimitIndices()[targetWrapDim] ==
+                  baseOperandType.getShape()[targetWrapDim] &&
+              sliceUser.getStrides()[targetWrapDim] == 1) {
+            candidates.push_back({wrapOfSlice, sliceUser});
+          }
+        }
+      }
+    }
+
+    if (candidates.size() <= 1) {
+      return rewriter.notifyMatchFailure(
+          triggerWrapOp,
+          "Rewrite condition not met (only found the trigger candidate)");
+    }
+
```

Comment:
@jumerckx can you apply the similar fix I made in https://github.com/EnzymeAD/Enzyme-JAX/commit/53652ee51583d80d5e016836e86e1ccc649a34ba